src/main/scala/strawman/collection/immutable/ImmutableArray.scala

```diff
@@ -0,0 +1,102 @@
+package strawman.collection.immutable
+
+import strawman.collection.mutable.ArrayBuffer
+import strawman.collection.{IterableFactory, IterableOnce, Iterator, View}
+
+import scala.{Any, Boolean, Int, Nothing}
+import scala.runtime.ScalaRunTime
+import scala.Predef.{???, intWrapper}
+
+/**
+  * An immutable array.
+  *
+  * Supports efficient indexed access and has a small memory footprint.
+  */
+class ImmutableArray[+A] private (private val elements: scala.Array[Any]) extends IndexedSeq[A] with SeqOps[A, ImmutableArray, ImmutableArray[A]] {
+
+  protected[this] def fromIterable[E](it: strawman.collection.Iterable[E]): ImmutableArray[E] = ImmutableArray.fromIterable(it)
+
+  protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[A]): ImmutableArray[A] = ImmutableArray.fromIterable(coll)
+
+  def length: Int = elements.length
+
+  override def knownSize: Int = elements.length
+
+  def apply(i: Int): A = ScalaRunTime.array_apply(elements, i).asInstanceOf[A]
+
+  def iterator(): Iterator[A] = view.iterator()
+
+  override def map[B](f: A => B): ImmutableArray[B] = ImmutableArray.tabulate(length)(i => f(apply(i)))
+
+  override def flatMap[B](f: A => IterableOnce[B]): ImmutableArray[B] = ImmutableArray.fromIterable(View.FlatMap(coll, f))
+
+  def :+ [B >: A](elem: B): ImmutableArray[B] = {
+    val dest = scala.Array.ofDim[Any](length + 1)
+    java.lang.System.arraycopy(elements, 0, dest, 0, length)
+    dest(length) = elem.asInstanceOf[Any]
+    new ImmutableArray(dest)
+  }
```

Comment:
Not copying every time would require a more elaborate copy-on-write implementation with implications for multi-threaded sharing. It would also lead to unpredictable performance in many cases and prevent use cases (like for varargs) where we need access to the underlying array without copying.

Such a data structure may be useful but it shouldn't be part of a simple `ImmutableArray`.